gobblin-core/src/main/java/gobblin/publisher/S3DataPublisher.java

```
@@ -0,0 +1,147 @@
+                      ConfigurationKeys.LOCAL_FS_URI));
+      this.fss.add(FileSystem.get(uri, conf));
+    }
+  }
+
+  /**
+   * Reads the S3_FILE_TO_PUBLISH key from the writer task, buffers them into one file,
+   * and dumps it all to S3.
+   *
+   * @param tasks the collection of tasks which we are publishing
+   * @throws java.io.IOException if the publish fails
+   */
+  @Override
+  public void publishData(Collection<? extends WorkUnitState> tasks) throws IOException {
+    InputStream is = null;
+    long contentLength = 0;
+    for (WorkUnitState task : tasks) {
+      for (int i = 0; i < branches; i++) {
+        String writerFilePathKey =
+                ForkOperatorUtils.getPropertyNameForBranch(ConfigurationKeys.WRITER_FILE_PATH, branches, i);
+        if (!task.contains(writerFilePathKey)) {
+          // Skip this branch as it does not have data output
+          continue;
+        }
+        Path writerOutputDir = new Path(task
+                .getProp(ForkOperatorUtils.getPropertyNameForBranch(ConfigurationKeys.WRITER_OUTPUT_DIR, branches, i)),
+                task.getProp(writerFilePathKey));
+        String fileName = task.getProp(ConfigurationKeys.S3_FILE_TO_PUBLISH);
+        Path writerFile = new Path(writerOutputDir, fileName);
+        contentLength += this.fss.get(i).getFileStatus(writerFile).getLen();
+
+        if (is == null) {
+          is = this.fss.get(i).open(writerFile);
+        } else {
+          is = new SequenceInputStream(this.fss.get(i).open(writerFile), is);
+        }
+        LOG.info("<" + writerFile + "> has been appended to s3 stream");
+      }
+    }
+    // get config parameters
+    String awsAccessKey = this.state.getProp(ConfigurationKeys.AWS_ACCESS_KEY);
+    String awsSecretKey = this.state.getProp(ConfigurationKeys.AWS_SECRET_KEY);
+    String bucket = this.state.getProp(ConfigurationKeys.S3_BUCKET_NAME);
```

should probably throw if not set?